include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +390,488 @@ class BidirectionalNode
   NeighborContainer Predecessors;
 };
 
+template<typename T>
+concept IsBidirectionalNode = requires {
+  T::is_bidirectional_node;
+  typename llvm::Inverse<T *>;
+};
+
+namespace detail {
+/// The parameters deciding specifics of the base type `NewNode` extends are
+/// non-trivial. That's why those decisiion were wrapped inside this struct
+/// to minimize clutter.
+///
+/// \note At this step `TheNode` has not been declared yet, thus we accept a
+///       template parameter that has the same signature as `TheNode` that
+///       will be declared later. This allows us to use it as if it was
+///       declared, provided that only the real `TheNode` is used as this
+///       argument.
+template<typename Node,
+         typename EdgeLabel,
+         bool HasParent,
+         size_t SmallSize,
+         template<typename, typename, bool, size_t, typename, size_t, bool>
+         class TheNode,
+         typename FinalType,
+         size_t ParentSmallSize,
+         bool ParentHasEntryNode>
+struct NewNodeBaseTCalc {
+  static constexpr bool
+    NoDerivation = std::is_same_v<FinalType, std::false_type>;
+  using NodeType = TheNode<Node,
+                           EdgeLabel,
+                           HasParent,
+                           SmallSize,
+                           FinalType,
+                           ParentSmallSize,
+                           ParentHasEntryNode>;
+  using DerivedType = std::conditional_t<NoDerivation, NodeType, FinalType>;
+  using GenericGraph = GenericGraph<DerivedType,
+                                    ParentSmallSize,
+                                    ParentHasEntryNode>;
+  using ParentType = Parent<GenericGraph, Node>;
+  using Result = std::conditional_t<HasParent, ParentType, Node>;
+};
+
+template<typename NodeType, typename LabelType>
+struct OwningEdge {
+  NodeType *Neighbor;
+  std::unique_ptr<LabelType> Label;
+};
+template<typename NodeType, typename LabelType>
+struct NonOwningEdge {
+  NodeType *Neighbor;
+  LabelType *Label;
+};
+template<typename NodeType, typename LabelType>
+struct EdgeView {
+  NodeType &Neighbor;
+  LabelType &Label;
+
+  explicit EdgeView(OwningEdge<NodeType, LabelType> &E) :
+    Neighbor(*E.Neighbor), Label(*E.Label) {}
+  explicit EdgeView(NonOwningEdge<NodeType, LabelType> &E) :
+    Neighbor(*E.Neighbor), Label(*E.Label) {}
+};
```

Comment:
Do we really need `EdgeView`? Can't we replace it with a simple `NonOwningEdge`?

Reply:
I don't really like the idea of exposing pointers to the user, when I can help it.
Technically, it's not that different, but reference-based interfaces just feel nicer. And any decent compiler will optimize extra stack allocations away, so that shouldn't be a problem.

Reply:
Are talking about the fact that `NonOwningEdge` has pointer fields? Can we switch to references there?
I'm trying to minimize the boilerplate/redundant code to keep things simple.

Reply:
For one, I prefer to keep `NonOwningEdge` mutable, there could probably be some potential optimization based on reusing existing edges.
Moreover, in my experience, reference members mess with copying-moving of objects. And since we're storing those in a dynamic container, that'd probably cause problems too.

Either way, I prefer using pointers for actual data structures and references for user interfaces, it makes distinguishing those easier and, in general, keeps interfaces clearer.

Reply:
If switching to `NonOwningEdge` enables us to drop 4 classes I'd consider it. But that's not probably the case, also due the discussion about `ConstEdgeView`.

Reply:
Well, we could probably combine those four classes into one using template-trickery but that's definitely not worth it. Also it'd drop readability even lower.